doc/design/block_execution.md

```diff
@@ -0,0 +1,147 @@
+  - multi-threads/multi-GPUs in a cluster: Each node in the cluster supports multi-threads/multi-GPUs.
+
+3. Model Parallelism
+  - Operators in a block can locate in different devices(CPU/GPU/FPGA) in a stand-alone machine. Users have to set device id for every operator manually.
+  - We should notice another situation where operators in a block can be executed in different CPU threads or CUDA streams to achieve high performace, but they are actually in the same device. It's scheduled by Paddle automatically, and users don't need to set specific CPU thread id or CUDA stream id. 
+
+4. Long-term
+  - Paddle schedules CPU/GPU/FPGA in a stand-alone machine automatically, makes full use of them to get best performance when executing a neural network.
+  - Paddle schedules CPU/GPU/FPGA in a cluster automatically, and overlaps communication and computation well to get best performance.
+
+  
+  
+## Analysis
+
+The execution of a neural network topology in PaddlePaddle is divided into two parts, complie-time and run-time.
+
+At complie-time, a ProgramDesc will be generated. At run-time, the ProgramDesc will be executed on specific hardwares.
+
+We can refer to the design of [computation-graphs](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/refactorization.md#computation-graphs).
+
+### Compile-time analysis
+
+In order to realize features mentioned in overview, we have to do several transform pass to ProgramDesc:
+
+- forward pass: generate OpDesc and VarDesc for forward operators
+
+- backward pass: generate OpDesc and VarDesc for backward operators
+
+- placement pass: generate OpDesc and VarDesc for send/recv/copy operators. Data Parallelism and Model Parallelism will be considered at this stage. We have a placement policy to set concrete device id for every operator. And send/recv/copy operators will be inserted at correct position. 
+
+- optimize pass: generate OpDesc and VarDesc for optimize operators
+
+Each pass will modify the global ProgramDesc.
+
+### Run-time analysis
+
+At run-time, the ProgramDesc generated at compile-time is actually executed on a specific hardware environment. `Executor` is introduced to execute a ProgramDesc. And `Executor` will have a `Run` interface.
+
+In PaddlePaddle, we use DeviceContext to manage hardware resources. There are CPUDeviceContext and CUDADeviceContext for CPU and GPU respectively.A CUDADeviceContext actually is associated with a specific CUDA stream. And Operator will run on a specfic DeviceContext. 
+
+However, users' configuration of hardware is simple, maybe only some GPU card ids will be given. In placement pass of compile-time, only source device id and destination device id will bet set.
+
+So, `Executor` need to have a `DeviceContextManager` to initialize some DeviceContext at the very beginning. `DeviceContextManager` maneges all DeviceContexts in all devices.
+
+The `Executor` is defined as follows:
+
+```
+class Executor {
+public:
+  Executor(ProgramDesc*);
```

Comment:
PaddlePaddle will `eval()` `ProgramDesc` very often, should we create a `Executor` every time a `ProgramDesc` is evaluated?

In my mind, one local session will create and own a single `Executor`, and `ProgramDesc` will be sent to the `Executor` to run.

Reply:
Eval will pass a ProgramDesc pointer to Executor. It consumes little time.

Reply:
Does Executor create the device contexts? If so, creating the Executor is not cheap.